packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx

```diff
@@ -108,16 +122,28 @@ function ControlPoints( {
 	onStartControlPointChange,
 	onStopControlPointChange,
 	__experimentalIsRenderedInSidebar,
-} ) {
-	const controlPointMoveState = useRef();
+}: ControlPointsProps ) {
+	const controlPointMoveState: React.MutableRefObject<
+		ControlPointMoveState | undefined
+	> = useRef();
 
-	const onMouseMove = ( event ) => {
+	const onMouseMove = ( event: MouseEvent ) => {
```

Comment:
From our pairing review:

```diff
diff --git a/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx b/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
index 3fddb545ba..e82f5ec79a 100644
--- a/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
+++ b/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
@@ -128,22 +128,28 @@ function ControlPoints( {
 	> = useRef();
 
 	const onMouseMove = ( event: MouseEvent ) => {
-		const relativePosition = getHorizontalRelativeGradientPosition(
-			event.clientX,
-			gradientPickerDomRef.current
-		);
-
-		const { initialPosition, index, significantMoveHappened } =
-			controlPointMoveState?.current ?? {};
+		// TODO: make sure that the "gradientPickerDomRef.current === null" makes sense
+		// ie. if gradientPickerDomRef.current === null, relativePosition would be
+		// `undefined`, which means that `updateControlPointPosition` would end up creating
+		// control points with an `undefined` position. We need to make sure that that's
+		// not a desirable state.
 		if (
 			controlPointMoveState.current === undefined ||
-			initialPosition === undefined ||
-			index === undefined ||
-			relativePosition === undefined
+			gradientPickerDomRef.current === null
 		) {
 			return;
 		}
 
+		// TODO: type getHorizontalRelativeGradientPosition so that it knows that
+		// its return value is `undefined` only if the second argument is null
+		const relativePosition = getHorizontalRelativeGradientPosition(
+			event.clientX,
+			gradientPickerDomRef.current
+		) as number;
+
+		const { initialPosition, index, significantMoveHappened } =
+			controlPointMoveState.current;
+
 		if (
 			! significantMoveHappened &&
 			Math.abs( initialPosition - relativePosition ) >=

```

Reply:
> make sure that the "gradientPickerDomRef.current === null" makes sense

I did some digging and it looks like the `gradientPickerDomRef.current === null` check is the right move to me - I can't see a case where control points with `undefined` position is desirable 👍 

> type getHorizontalRelativeGradientPosition so that it knows that its return value is `undefined` only if the second argument is null

I tried using a conditional type here, but couldn't get it working so that TypeScript could infer the correct return type.
<details>

<summary>I've experimented with an override instead:</summary>

```diff

diff --git a/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx b/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
index 9668022c62..2f07b6e3b4 100644
--- a/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
+++ b/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
@@ -136,7 +136,7 @@ function ControlPoints( {
 		const relativePosition = getHorizontalRelativeGradientPosition(
 			event.clientX,
 			gradientPickerDomRef.current
-		) as number;
+		);
 
 		const { initialPosition, index, significantMoveHappened } =
 			controlPointMoveState.current;
diff --git a/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts b/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
index 271d4fbc16..f82aaac4f0 100644
--- a/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
+++ b/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
@@ -93,6 +93,17 @@ export function updateControlPointColorByPosition(
 	return updateControlPointColor( points, index, newColor );
 }
 
+// `getHorizontalRelativeGradientPosition` should only return `undefined` if
+// the `containerElement` argument is `null` (i.e. when the relevant ref is
+// first created, but before the element is rendered and assigned to the ref).
+export function getHorizontalRelativeGradientPosition(
+	mouseXcoordinate: number,
+	containerElement: HTMLDivElement
+): number;
+export function getHorizontalRelativeGradientPosition(
+	mouseXcoordinate: number,
+	containerElement: null
+): undefined;
 export function getHorizontalRelativeGradientPosition(
 	mouseXCoordinate: number,
 	containerElement: HTMLDivElement | null


```

</details>

But I'm not sure if that's the right way to go.

Reply:
Done!

Reply:
Additional note - when we looked at this change earlier, you floated the idea of a third overload that would accept and return a union of the two possible types, but we didn't feel it was needed. It turns out it is, otherwise we get an error [when inserting a control point](https://github.com/WordPress/gutenberg/blob/9e2185782f1f913a5f9b49c68fab9312d3cf3bb0/packages/components/src/custom-gradient-picker/gradient-bar/index.tsx#L101).

Apparently if the call could accept a union, but none of the overrides do, it errors... even if each union member matches an available override. I did a little hunting and [it looks like this is intentional](https://github.com/WordPress/gutenberg/blob/9e2185782f1f913a5f9b49c68fab9312d3cf3bb0/packages/components/src/custom-gradient-picker/gradient-bar/index.tsx#L101), so I've added the third override for the union.